File: libs/sdk-python/src/daytona/common/code_interpreter.py
# Copyright 2025 Daytona Platforms Inc.
# SPDX-License-Identifier: Apache-2.0

from typing import Any, Awaitable, Callable, Optional, TypeVar, Union

from pydantic import BaseModel


class OutputMessage(BaseModel):
    """Represents stdout or stderr output from code execution.

    Attributes:
        output: The output content.
    """

    output: str


class ExecutionError(BaseModel):
    """Represents an error that occurred during code execution.

    Attributes:
        name: The error type/class name (e.g., "ValueError", "SyntaxError").
        value: The error value.
        traceback: Full traceback of the error.
    """

    name: str
    value: str
    traceback: str = ""


class ExecutionResult(BaseModel):
    """Result of code execution.

    Attributes:
        stdout: Standard output from the code execution.
        stderr: Standard error output from the code execution.
        error: Error details if execution failed, None otherwise.
    """

    stdout: str = ""
    stderr: str = ""
    error: Optional[ExecutionError] = None


# Type aliases for callbacks
T = TypeVar("T")
OutputHandler = Union[
    Callable[[T], Any],
    Callable[[T], Awaitable[Any]],
]

File: libs/sdk-ruby/lib/daytona/common/code_interpreter.rb
# frozen_string_literal: true

module Daytona
  # Represents stdout or stderr output from code execution
  class OutputMessage
    # @return [String] The output content
    attr_reader :output

    # @param output [String]
    def initialize(output:)
      @output = output
    end
  end

  # Represents an error that occurred during code execution
  class ExecutionError
    # @return [String] The error type/class name (e.g., "ValueError", "SyntaxError")
    attr_reader :name

    # @return [String] The error value
    attr_reader :value

    # @return [String] Full traceback of the error
    attr_reader :traceback

    # @param name [String]
    # @param value [String]
    # @param traceback [String]
    def initialize(name:, value:, traceback: '')
      @name = name
      @value = value
      @traceback = traceback
    end
  end

  # Result of code execution
  class ExecutionResult
    # @return [String] Standard output from the code execution
    attr_accessor :stdout

    # @return [String] Standard error output from the code execution
    attr_accessor :stderr

    # @return [ExecutionError, nil] Error details if execution failed, nil otherwise
    attr_accessor :error

    def initialize(stdout: '', stderr: '', error: nil)
      @stdout = stdout
      @stderr = stderr
      @error = error
    end
  end
end
